Fix CoAP option length checking and drop COAP_EXTENDED_OPTIONS_LEN from LwM2M

mike-scott opened the pull request:
Currently, we check the length of an option value in the coap_packet_append_option() function.  This isn't required as we're appending to a net_pkt and not using struct coap_option where the limitation is imposed. Instead, we should check the option value length in parse_option() where we assign the value to a struct coap_option.

Also, when moving to the new CoAP API, I thought we would need to parse incoming option values longer than 12 characters. This hasn't proven to be true, so let's remove the auto-selection of this config.  If needed user can set this option later.

Commit:
net: coap: Fix error handling for coap option length

Currently, we check the length of an option value in the
coap_packet_append_option() function.  This isn't required as
we're appending to a net_pkt and not using struct coap_option
where the limitation is imposed.

Instead, we should check the option value length in
parse_option() where we assign the value to a struct
coap_option.

Signed-off-by: Michael Scott <michael.scott@linaro.org>

Commit:
net: lwm2m: remove auto select of COAP_EXTENDED_OPTIONS_LEN config

When moving to the new CoAP API, I thought we would need to parse
incoming option values longer than 12 characters.

This hasn't proven to be true, so let's remove the auto-selection of
this config.  If needed user can set this option later.

Signed-off-by: Michael Scott <michael.scott@linaro.org>

mike-scott commented:
Rebased on current master to make sure it applies correctly.

rbtchc reviewed (approved):
LGTM

rveerama1 commented on subsys/net/lib/coap/coap.c:
```diff
@@ -217,6 +217,14 @@ static int parse_option(const struct coap_packet *cpkt,
 
 	if (option) {
+		if (len > sizeof(option->value)) {
+			NET_ERR("Incoming COAP option length is %u!\n"
```
suggestion: what about single line error message and add more comments in the code? Two advantages with this approach 1) User can easily grep "error message" and 2) Less binary size.

mike-scott replied:
Sounds good.  How about a comment like this above the if check:
/*
 * Make sure the option data will fit into the value field of coap_option.
 * NOTE: To expand the size of the value field set:
 * CONFIG_COAP_EXTENDED_OPTIONS_LEN=y
 * CONFIG_COAP_EXTENDED_OPTIONS_LEN_VALUE=<new size>
 */

Then error can be something like:
"%u is > coap_option->value size(%u)!"

rveerama1 replied:
Ok, good.

nashif commented on subsys/net/lib/coap/coap.c:
```diff
@@ -217,6 +217,14 @@ static int parse_option(const struct coap_packet *cpkt,
 	}
 
 	if (option) {
```
how about one if statement here instead of two?

mike-scott replied:
We need 2 checks because because we're looking at 3 possible scenarios here:
- It's not an error condition if this function gets an option parameter == NULL.  So we look for this condition.  This happens when a user passes in an array of coap_options to parse_options() and we run out of indexes before the end of the option data.  [NOTE: This *could* generate a warning of some sort as well.]
- The function has a valid option pointer, but the length of the value field isn't large enough to hold the incoming data (This is the error I'm looking for -- and something the developer should be made aware of because it's fixable.)
- The function has a valid option pointer, and the length is ok to use it.

mike-scott commented:
Updated with V2.  Diff: https://hastebin.com/ocakeciwec.cs

Changes:
- Moved commentary from error message to comment
- Single line error message
- Changed error from EINVAL to ENOMEM

jukkar reviewed (approved):
LGTM

mike-scott commented:
@rveerama1 @nashif any further comments on this?

rveerama1 reviewed (approved):
LGTM
